libr/core/cmd.c

```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
 	return res;
 }
 
+DEFINE_HANDLE_TS_FCN(foreach_addrsize_command) {
+	RCore *core = state->core;
+	TSNode command = ts_node_named_child (node, 0);
+	uint32_t i = 1;
+	bool ret = true;
+	TSNode seek_addr_node = ts_node_named_child (node, i);
+	TSNode blk_sz_node = ts_node_named_child (node, i + 1);
+	while (!ts_node_is_null (seek_addr_node)) {
+		char *seek_addr_str = ts_node_handle_arg (state, node, seek_addr_node, i);
+		char *blk_sz_str = ts_node_handle_arg (state, node, blk_sz_node, i + 1);
+		ut64 seek_addr_val = r_num_get (NULL, seek_addr_str);
+		ut64 blk_sz_val = r_num_get (NULL, blk_sz_str);
+		free (seek_addr_str);
+		free (blk_sz_str);
+		r_core_seek (core, seek_addr_val, 1);
+		r_core_block_size (core, blk_sz_val);
+		ret &= handle_ts_command (state, command);
+		i += 2;
+		seek_addr_node = ts_node_named_child (node, i);
+		blk_sz_node = ts_node_named_child (node, i + 1);
+	}
+	return ret;
+}
+
+DEFINE_HANDLE_TS_FCN(foreach_bb_command) {
+	RCore *core = state->core;
+	TSNode command = ts_node_named_child (node, 0);
+	RAnalFunction *fcn = r_anal_get_fcn_in (core->anal, core->offset, 0);
+	ut64 offorig = core->offset;
+	ut64 obs = core->blocksize;
+	if (!fcn) {
+		return false;
+	}
+
+	RListIter *iter;
+	RAnalBlock *bb;
+	bool ret = true;
+	r_list_foreach (fcn->bbs, iter, bb) {
+		r_core_seek (core, bb->addr, 1);
+		r_core_block_size (core, bb->size);
+		ret &= handle_ts_command (state, command);
+	}
+	r_core_block_size (core, obs);
+	r_core_seek (core, offorig, 1);
+	return ret;
+}
+
+DEFINE_HANDLE_TS_FCN(foreach_cmd_command) {
+	// convert @@@c: command into a @@@= one, by using the output of the
+	// in_cmd as addr/blksz of @@@=
+	TSNode in_cmd = ts_node_named_child (node, 1);
+	substitute_args_init (state, node);
+
+	RList *edits = r_list_newf ((RListFree)free_tsr2cmd_edit);
+
+	char *in_cmd_out = do_handle_substitution_cmd (state, in_cmd);
+	in_cmd_out = escape_special_chars (in_cmd_out, SPECIAL_CHARS_REGULAR);
+	struct tsr2cmd_edit *e = create_cmd_edit (state, in_cmd, in_cmd_out);
+	r_list_append (edits, e);
+
+	TSNode op = ts_node_child (node, 1);
+	e = create_cmd_edit (state, op, strdup ("@@@="));
```

Comment:
passing strdup? isnt it leaking imho it should take a const char *

Reply:
It does not as `e` is then freed by `free_tsr2cmd_edit`, which expects the string to be free-able.